### PR TITLE
Enforce expected hash presence when creating and assigning references

### PR DESCRIPTION
**Pull request by adutra**
This commit also refactors the HashValidator and HashResolver APIs
to achieve better error messages.

This commit also enhances the set of tests in `AbstractRelativeReferences` and adds:

* Tests for references without hashes (resolving to HEAD when allowed)
* Tests for all kinds of references (branches, tags and detached)
* Tests for expected hashes in various situations: 
  * Resolving to HEAD
  * Resolving to an ancestor of HEAD
  * Resolving to an unreachable commit from HEAD
  * Non-existent hash

Some validation constraints in `TreeService` and `Merge` related to hashes were
lifted, in order to let HashValidator validate them and produce
standardized error messages.

### CODE REVIEW

File: servers/services/src/main/java/org/projectnessie/services/hash/HashResolver.java [hunk 1]
```diff
@@ -94,12 +97,8 @@ public ResolvedHash resolveHashOnRef(
         ReferenceInfo<CommitMeta> refInfo = store.getNamedRef(namedRef, GetNamedRefsParams.DEFAULT);
         ref = refInfo.getNamedRef();
         currentHead = refInfo.getHash();
-        // Shortcut: use HEAD, if hashOnRef is null or empty
-        if (hashOnRef == null || hashOnRef.isEmpty()) {
```
Comment (adutra) [inline servers/services/src/main/java/org/projectnessie/services/hash/HashResolver.java:98]: This shortcut was bypassing the validation rules so I had to get rid of it.

File: servers/services/src/main/java/org/projectnessie/services/hash/HashValidator.java [hunk 2]
```diff
@@ -17,50 +17,122 @@
...
 
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
 import javax.annotation.Nullable;
+import org.projectnessie.versioned.BranchName;
+import org.projectnessie.versioned.NamedRef;
+import org.projectnessie.versioned.TagName;
 
-@FunctionalInterface
-public interface HashValidator {
+public final class HashValidator {
```
Comment (adutra) [inline servers/services/src/main/java/org/projectnessie/services/hash/HashValidator.java:26]: The refactoring is mainly motivated by the need to preserve the field names when generating error messages. Otherwise it's functionally equivalent.

File: servers/services/src/main/java/org/projectnessie/services/hash/ResolvedHash.java [hunk 3]
```diff
@@ -36,29 +34,13 @@ public interface ResolvedHash extends WithHash<NamedRef> {
...
-  /**
-   * The user-provided hash. Only present if a hash was actually provided and successfully parsed;
-   * will be empty if the hash was implicitly resolved to the ref's HEAD.
-   *
-   * <p>If this is empty, then {@link #getHead()} is guaranteed to be non-empty.
-   *
-   * <p>If the provided hash had relative specs, this hash will be the result of resolving those
-   * specs against its absolute part, or the ref's HEAD if it doesn't have an absolute part.
-   */
-  Optional<Hash> getProvidedHash();
```
Comment (adutra) [inline servers/services/src/main/java/org/projectnessie/services/hash/ResolvedHash.java:48]: Not needed anymore.

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java [hunk 4]
```diff
@@ -308,16 +305,22 @@ public Reference assignReference(
...
+                      .refMustBeBranchOrTag()
+                      .hashMustBePresent()
+                      .hashMustNotBeAmbiguous());
       ResolvedHash newRef =
           getHashResolver()
               .resolveHashOnRef(
-                  assignTo.getName(), assignTo.getHash(), "Target hash", UNAMBIGUOUS_HASH);
+                  assignTo.getName(),
+                  assignTo.getHash(),
+                  new HashValidator("Target hash").hashMustBePresent().hashMustNotBeAmbiguous());
```
Comment (adutra) [inline servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java:323]: And this is the core fix for `assignReference`.

File: servers/services/src/main/java/org/projectnessie/services/spi/TreeService.java [hunk 5]
```diff
@@ -198,8 +198,6 @@ MergeResponse transplantCommitsIntoBranch(
           String branchName,
       @Valid
           @jakarta.validation.Valid
-          @NotNull
-          @jakarta.validation.constraints.NotNull
```
Comment (adutra) [inline servers/services/src/main/java/org/projectnessie/services/spi/TreeService.java:202]: It's better to let the hash validator generate the error messages, they will be more homogenous and more informational.

File: CHANGELOG.md [hunk 6]
```diff
@@ -20,6 +20,7 @@ as necessary. Empty sections will not end in the release notes.
 ### Changes
 - Java client API to assign/delete reference operations without specifying a concrete reference type
   (no change to REST API).
+- Creating and assigning references now require a target hash to be specified.
```
Comment (snazy) [inline CHANGELOG.md:23]:
```suggestion
- Creating and assigning references now requires a target hash to be specified.
```

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java [hunk 7]
```diff
@@ -330,7 +333,7 @@ public Reference assignReference(
           .canAssignRefToHash(oldRef.getNamedRef())
           .checkAndThrow();
 
-      getStore().assign(oldRef.getNamedRef(), oldRef.getProvidedHash(), newRef.getHash());
+      getStore().assign(oldRef.getNamedRef(), Optional.of(oldRef.getHash()), newRef.getHash());
```
Comment (snazy) [inline servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java:336]: I suspect, with this change, the `Optional` in the signatures of `VersionStore.assign` + `VersionStore.create` can go away.

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java [hunk 8]
```diff
@@ -87,13 +89,16 @@ void createFixtures() {
...
+                                 ^
+                               tag1
     */
     assumeTrue(outer.isNewModel());
     Branch base = outer.createBranchV2("base");
     base = outer.commitV2(base, key1, table1);
     c1 = base.getHash();
     base = outer.commitV2(base, key2, table2);
     c2 = base.getHash();
+    outer.createTagV2("tag0", base);
```
Comment (snazy) [inline servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java:101]: Looks like `tag0` is missing in the diagram above (`tag1`'s added).

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java [hunk 9]
```diff
@@ -123,143 +129,331 @@ public void createReferenceUnambiguous() {
...
   /**
-   * Can create a reference from a hash that is unambiguous and DETACHED. Expect a 200 response with
-   * the created reference.
+   * Cannot create a reference from a branch + hash that is ambiguous (writing operation). The hash
+   * must contain a starting commit ID. Expect a BAD_REQUEST error.
    */
   @Test
   @NessieApiVersions(versions = NessieApiVersion.V2)
-  public void createReferenceUnambiguousDetached() {
+  public void createReferenceFromBranchAmbiguous() {
```
Comment (snazy) [inline servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java:137]: Wonder if we need these tests as "pure REST tests" instead of using the Nessie API. I think, it's better to use the Nessie API and only have "pure REST tests" for test cases that do not work via the API.

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java (hunk 8 above)
Comment (adutra) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java]: Actually we won't need `tag0` anymore, I forgot to remove it.

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java (hunk 9 above)
Comment (adutra) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java:136]: I actually went for REST tests precisely because we can test any test case by forging the HTTP request by hand. Some of the test cases would indeed be untestable through the API.

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java (hunk 7 above)
Comment (adutra) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java:336]: Yes. Do you prefer that I make the change now, or in a follow-up PR?

File: api/model/src/main/java/org/projectnessie/api/v2/params/Merge.java [hunk 10]
```diff
@@ -93,14 +92,14 @@ public interface Merge extends BaseMergeTransplant {
...
-  @NotBlank
-  @jakarta.validation.constraints.NotBlank
   @Pattern(
       regexp = Validation.HASH_OR_RELATIVE_COMMIT_SPEC_REGEX,
       message = Validation.HASH_OR_RELATIVE_COMMIT_SPEC_MESSAGE)
   @jakarta.validation.constraints.Pattern(
       regexp = Validation.HASH_OR_RELATIVE_COMMIT_SPEC_REGEX,
       message = Validation.HASH_OR_RELATIVE_COMMIT_SPEC_MESSAGE)
+  @Nullable
+  @jakarta.annotation.Nullable
```
Comment (adutra) [inline api/model/src/main/java/org/projectnessie/api/v2/params/Merge.java:102]: Same here: goal is to let `HashValidator` produce the error message.

**Comment by adutra** [comment]
NesQuEIT test are passing now.

File: servers/services/src/main/java/org/projectnessie/services/hash/ResolvedHash.java [hunk 11]
```diff
@@ -36,29 +34,13 @@ public interface ResolvedHash extends WithHash<NamedRef> {
...
-   *
-   * <p>If this is empty, then {@link #getProvidedHash()} is guaranteed to be non-empty.
    */
   Optional<Hash> getHead();
 
-  /**
-   * The effective resolved hash, never {@code null}. Will either be {@link #getProvidedHash()} if
-   * present, or {@link #getHead()} otherwise.
-   */
+  /** The effective resolved hash, never {@code null}. */
```
Comment (dimas-b) [inline servers/services/src/main/java/org/projectnessie/services/hash/ResolvedHash.java:43]: nit: why not annotate as `@Nonnull`?

File: servers/services/src/main/java/org/projectnessie/services/hash/HashValidator.java [hunk 12]
```diff
@@ -17,50 +17,94 @@
...
    */
-  default HashValidator and(HashValidator other) {
-    return (name, parsed) -> {
-      validate(name, parsed);
-      other.validate(name, parsed);
-    };
+  @CanIgnoreReturnValue
+  public HashValidator hashMustNotBeAmbiguous() {
+    validator =
+        validator.andThen(
```
Comment (dimas-b) [inline servers/services/src/main/java/org/projectnessie/services/hash/HashValidator.java:102]: It looks like top-level services use this method and `HashResolver` can also call this, which will cause duplicate conditions. Operation will still be correct, but it feels awkward... Would it be simpler to just keep several validation flags?

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java [hunk 13]
```diff
@@ -308,16 +305,22 @@ public Reference assignReference(
...
     try {
-      checkArgument(expectedHash != null, "Expected hash must be provided");
 
       ResolvedHash oldRef =
           getHashResolver()
               .resolveHashOnRef(
-                  referenceName, expectedHash, "Expected hash", REQUIRED_UNAMBIGUOUS_HASH);
+                  referenceName,
+                  expectedHash,
+                  new HashValidator("Reference to assign from", "Expected hash")
```
Comment (dimas-b) [inline servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java:314]: maybe just `Reference to assign`?.. not sure I get the meaning of `from` in the message :thinking:

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java (hunk 7 above)
Comment (dimas-b) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java:336]: Let's remove `Optional` now, because this PR make the hash required and `Optional` unnecessary.

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java [hunk 14]
```diff
@@ -710,11 +718,19 @@ public MergeResponse mergeRefIntoBranch(
 
       ResolvedHash fromRef =
           getHashResolver()
-              .resolveHashOnRef(fromRefName, fromHash, "Source hash", REQUIRED_UNAMBIGUOUS_HASH);
+              .resolveHashOnRef(
+                  fromRefName,
+                  fromHash,
+                  new HashValidator("Source hash").hashMustBePresent().hashMustNotBeAmbiguous());
```
Comment (dimas-b) [inline servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java:724]: nit: maybe add a helper method like `nonAmbiguousHash(String fieldname)`?

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java (hunk 9 above)
Comment (dimas-b) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java:136]: I agree. The java client API seems weaker than the REST API in terms of expressing relative references. It might be worth adding specific methods for "merge parent #N", "ancestor #Y", etc., where hashes are manipulated.

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java [hunk 15]
```diff
@@ -803,11 +1515,16 @@ void getDiffToRefAmbiguousDetached() {
     checkError(error, BAD_REQUEST, "To hash must contain a starting commit ID.");
```
Comment (dimas-b) [inline servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java:1515]: Would it make sense to add quotation marks around `To`, i.e. `"To" hash ...`? In current form the whole message is not very readable, IMHO.

This is not really related to the purpose of this PR, but we're changing how validation messages are constructed :thinking:

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java (hunk 9 above)
Comment (snazy) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java:136]: > Some of the test cases would indeed be untestable through the API.

fair

File: servers/services/src/main/java/org/projectnessie/services/hash/ResolvedHash.java (hunk 11 above)
Comment (adutra) [reply servers/services/src/main/java/org/projectnessie/services/hash/ResolvedHash.java:43]: The Immutables library has no support for `@Nonnull`: it encourages instead to consider everything as non-null and annotate the exceptions with `@Nullable` or use `Optional`. See http://immutables.github.io/immutable.html#nullable-attributes.

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java (hunk 14 above)
Comment (adutra) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java:724]: Now that `hashMustBePresent` has been merged with `hashMustNotBeAmbiguous` maybe that's not required anymore?

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java (hunk 7 above)
Comment (adutra) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java:335]: Done. We can also make a case for removing the `Optional` in:

* `VersionStore.commit.referenceHash`
* `MergeOp.expectedHash`
* `TransplantOp.expectedHash`

But if everybody agrees, I will look into that in another PR :-)

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java (hunk 15 above)
Comment (adutra) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java]: I used `"To" hash` as suggested... if that's not enough, we could maybe use `Hash to start the diff from ...` wdyt?

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java (hunk 13 above)
Comment (dimas-b) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java]: `Assignment target` SGTM

**Review by dimas-b — APPROVED** [review]
LGTM :+1: thx! :)
